File: evals/behavioral-analysis/data/csharp/template-injection/FStringInjectionDynamicEval.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: RazorEngineCore compiles caller Razor like Python eval(f"...") / f-string sinks

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text.Json;
using ModelContextProtocol.Server;
using RazorEngineCore;

namespace EvalSamples.TemplateInjection;

[McpServerToolType]
public static class FStringInjectionDynamicEval
{
    [McpServerTool]
    [Description("Format message with user variables and dynamic expressions.")]
    public static string FormatMessage(
        string template,
        string name = "",
        string custom_vars = "",
        bool enable_expressions = false,
        bool include_builtins = false)
    {
        var dict = new Dictionary<string, object?>
        {
            ["name"] = name,
            ["user"] = name,
            ["greeting"] = "Hello",
        };

        if (!string.IsNullOrEmpty(custom_vars))
        {
            try
            {
                using var doc = JsonDocument.Parse(custom_vars);
                foreach (var prop in doc.RootElement.EnumerateObject())
                    dict[prop.Name] = JsonSerializer.Deserialize<object>(prop.Value.GetRawText());
            }
            catch
            {
                /* ignore */
            }
        }

        if (include_builtins)
            dict["builtins_echo"] = true;

        object model = dict;

        var engine = new RazorEngine();
        try
        {
            var compiled = engine.Compile(template);
            var result = compiled.Run(model);
            return $"Formatted message: {result}";
        }
        catch (Exception ex)
        {
            return $"Format error: {ex.Message}";
        }
    }
}

File: evals/behavioral-analysis/data/rust/template-injection/fstring_injection_dynamic_eval.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Caller-controlled templates (Python f-string eval analogue) via Tera one_off.

use rmcp::{
    handler::server::tool::ToolRouter,
    tool, tool_handler, tool_router,
    ServerHandler,
    ServiceExt,
    transport::stdio,
};
use rmcp::model::{CallToolResult, Content};
use schemars::JsonSchema;
use serde::Deserialize;
use serde_json::Value;
use tera::{Context, Tera};

#[derive(Deserialize, JsonSchema)]
struct FmtArgs {
    template: String,
    #[serde(default)] name: String,
    #[serde(default)] custom_vars: String,
    #[serde(default)] enable_expressions: bool,
    #[serde(default)] include_builtins: bool,
}

fn merge_custom(base: serde_json::Map<String, Value>, raw: &str) -> serde_json::Map<String, Value> {
    let mut m = base;
    if raw.is_empty() {
        return m;
    }
    if let Ok(v) = serde_json::from_str::<Value>(raw) {
        if let Value::Object(obj) = v {
            for (k, v) in obj {
                m.insert(k, v);
            }
        }
    }
    m
}

#[derive(Clone)]
struct FStringServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl FStringServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "Format message with user variables and dynamic expressions.")]
    async fn format_message(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<FmtArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let mut obj = serde_json::Map::new();
        obj.insert("name".to_string(), Value::String(args.name.clone()));
        obj.insert("user".to_string(), Value::String(args.name.clone()));
        obj.insert("greeting".to_string(), Value::String("Hello".to_string()));
        obj = merge_custom(obj, &args.custom_vars);

        let mut ctx = Context::new();
        for (k, v) in &obj {
            ctx.insert(k, v);
        }
        if args.include_builtins {
            ctx.insert("_builtins", &"hint");
        }

        let _ = args.enable_expressions;

        let out = Tera::one_off(&args.template, &ctx, false)
            .unwrap_or_else(|e| format!("Format error: {e}"));

        Ok(CallToolResult::success(vec![Content::text(format!("Formatted message: {out}"))]))
    }
}

#[tool_handler]
impl ServerHandler for FStringServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = FStringServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
